File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.py
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.max_sum = float('-inf')
        
        def get_max_gain(node):
            if not node:
                return 0

            leftSum = get_max_gain(node.left)
            rightSum = get_max_gain(node.right)
            
            maxLeftSum = max(leftSum, 0)
            maxrightSum = max(rightSum, 0)
            
            current_path_sum = node.val + maxLeftSum + maxrightSum
            self.max_sum = max(self.max_sum, current_path_sum)
            
            return node.val + max(maxLeftSum, maxrightSum)

        get_max_gain(root)
        return self.max_sum

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {

        int[] arr = new int[2];
        arr[0] = k;
        inOrder(root, arr);
        return arr[1];
    }

    public void inOrder(TreeNode root, int[] arr) {
        if (root == null) {
            return;
        }
        inOrder(root.left, arr);
        arr[0] -= 1;
        if (arr[0] == 0) {
            arr[1] = root.val;
        }
        inOrder(root.right, arr);
    }
}
